### PR TITLE
RV32M1 SoC and VEGAboard support

### PR DESCRIPTION
**Pull request by mbolivar**
Edit: removed DNM tag, the series this depended on was merged

This adds support for the OpenISA VEGAboard's RI5CY processor, as well as board support for VEGAboard itself.

It does add a new vendor HAL; it's BSD-3-Clause licensed.

### CODE REVIEW

**Comment by mbolivar** [comment]
Fixed some of the style problems. Working on getting a license and copyright for openocd.cfg.

Many of the checkpatch suggestions would break if applied, so will have to be ignored.

File: arch/common/gen_isr_tables.py
Comment (SebastianBoe) [inline arch/common/gen_isr_tables.py]: level2_offset.get("CONFIG_2ND_LVL_ISR_TBL_OFFSET", None)

File: arch/riscv32/core/fatal.c
Comment (SebastianBoe) [inline arch/riscv32/core/fatal.c]: Users don't expect to see ifdef's around CONFIG_PRINTK.

We should instead drop the #ifdef CONFIG_PRINTK around mcause. This has no performance impact, and makes this function easier to read.

File: cmake/compiler/gcc.cmake
Comment (SebastianBoe) [inline cmake/compiler/gcc.cmake]: list(APPEND

A part from that, this patch looks sane.

File: ext/hal/openisa/vega_sdk_riscv/Kconfig
Comment (SebastianBoe) [inline ext/hal/openisa/vega_sdk_riscv/Kconfig]: "HAS_" options should not have any direct effect on the build when enabled.

It seems you have no need for a "HAS_" option, so you can simply rename this to "VEGA_SDK_HAL".

File: soc/riscv32/openisa_rv32m1/CMakeLists.txt
Comment (SebastianBoe) [inline soc/riscv32/openisa_rv32m1/CMakeLists.txt]: This assertion should be moved into Kconfig.

Also, it doesn't make sense to have code after a FATAL_ERROR.

File: soc/riscv32/openisa_rv32m1/linker.ld
Comment (SebastianBoe) [inline soc/riscv32/openisa_rv32m1/linker.ld]: It is OK to have these assertions here. But they should redundantly exist in Kconfig as well.

**Review by SebastianBoe — CHANGES_REQUESTED** [review]
Generally this looks very good.

But see misc. review comments.

File: arch/riscv32/Kconfig
Comment (dbkinder) [inline arch/riscv32/Kconfig]: I think these underscores are causing the doc failure (WARNING: Inline emphasis start-string without end-string), so change to ``\_\_soc_save_context and \_\_soc_restore_context``

File: arch/riscv32/Kconfig
Comment (dbkinder) [inline arch/riscv32/Kconfig]: These underscores are OK (in a code block)

File: boards/riscv32/rv32m1_vega/doc/rv32m1_vega.rst
Comment (dbkinder) [inline boards/riscv32/rv32m1_vega/doc/rv32m1_vega.rst]: Our project conventions are to use dashes ``----`` for this level of heading

File: boards/riscv32/rv32m1_vega/doc/rv32m1_vega.rst
Comment (dbkinder) [inline boards/riscv32/rv32m1_vega/doc/rv32m1_vega.rst]: ditto (dashes here too)

File: boards/riscv32/rv32m1_vega/doc/rv32m1_vega.rst
Comment (dbkinder) [inline boards/riscv32/rv32m1_vega/doc/rv32m1_vega.rst]: Recommend breaking this long line into multiple lines (ending with a ``\``)

**Review by dbkinder — CHANGES_REQUESTED** [review]
Nice documentation, thanks!

File: arch/riscv32/core/fatal.c
Comment (mbolivar) [reply arch/riscv32/core/fatal.c]: This was merged as part of #11200. I will add a commit making this change to this series.

File: cmake/compiler/gcc.cmake
Comment (mbolivar) [reply cmake/compiler/gcc.cmake]: Much better.

File: drivers/timer/rv32m1_lptmr_timer.c
Comment (mbolivar) [reply drivers/timer/rv32m1_lptmr_timer.c]: Ah, of course! Thanks.

File: boards/riscv32/rv32m1_vega/doc/rv32m1_vega.rst
Comment (mbolivar) [reply boards/riscv32/rv32m1_vega/doc/rv32m1_vega.rst]: Are you sure? I get:

```
/home/mbolivar/src/zephyr/doc/_build/rst/doc/boards/riscv32/rv32m1_vega/doc/rv32m1_vega.rst:310: WARNING: Title level inconsistent:

Option 1 (Recommended): Prebuilt Toolchain and OpenOCD
------------------------------------------------------
/home/mbolivar/src/zephyr/doc/_build/rst/doc/boards/riscv32/rv32m1_vega/doc/rv32m1_vega.rst:372: WARNING: Title level inconsistent:

Option 2: Building Toolchain and OpenOCD From Source
----------------------------------------------------
```

**Comment by mbolivar** [comment]
@dbkinder @SebastianBoe comments addressed, thanks! Let's see what CI says.

**Review by dbkinder — CHANGES_REQUESTED** [review]

**Review by SebastianBoe — APPROVED** [review]

**Comment by alistair23** [comment]
Can this be implemented following the RISC-V spec and not have a dependency on custom pulp toolchains?

With this diff I am able to build it using upstream GCC (not yet tested on hardware)

```
diff --git a/soc/riscv32/openisa_rv32m1/CMakeLists.txt b/soc/riscv32/openisa_rv32m1/CMakeLists.txt
index e4b5a8d535..d79089d3c4 100644
--- a/soc/riscv32/openisa_rv32m1/CMakeLists.txt
+++ b/soc/riscv32/openisa_rv32m1/CMakeLists.txt
@@ -3,7 +3,7 @@
 # SPDX-License-Identifier: Apache-2.0
 
 if(CONFIG_SOC_OPENISA_RV32M1_RI5CY)
-  zephyr_compile_options(-march=rv32imcxpulpv2)
+  zephyr_compile_options(-march=rv32imfdc)
 elseif(CONFIG_SOC_OPENISA_RV32M1_ZERO_RISCY)
   zephyr_compile_options(-march=rv32imc)
 endif()
```

File: boards/riscv32/rv32m1_vega/doc/rv32m1_vega.rst
Comment (alistair23) [inline boards/riscv32/rv32m1_vega/doc/rv32m1_vega.rst]: Is this really the only way? Anyone who wants to use this board needs to spend hundreds of dollars on a JTAG debugger?

File: boards/riscv32/rv32m1_vega/doc/rv32m1_vega.rst
Comment (dbkinder) [inline boards/riscv32/rv32m1_vega/doc/rv32m1_vega.rst]: Please use our document conventions and use ``------`` here and below instead of ``~~~~~~`` for this heading level.

**Review by dbkinder — CHANGES_REQUESTED** [review]

File: boards/riscv32/rv32m1_vega/doc/rv32m1_vega.rst
Comment (mbolivar) [reply boards/riscv32/rv32m1_vega/doc/rv32m1_vega.rst]: > The [open-isa website](http://open-isa.org/get-started/) implies that everything you need is included in the box. Is there no other way to use the board besides buying a specialised JTAG debugger?

There is no other way.

**Comment by mbolivar** [comment]
@dbkinder sorry, neglected to mention, I've addressed your comments.

@alistair23: 

> With this diff I am able to build it using upstream GCC (not yet tested on hardware), if this could at least be described somewhere that would help others.

That's not really in my purview. I'm just a contractor and the recommendation I got from OpenISA is to use their toolchain. At any rate, I don't believe your patch is correct; specifically this part:

```
  zephyr_compile_options(-march=rv32imfdc)
```

If you check the reference manual, the core is an IMC core, but it's not documented to have any floating point support (the F and D extensions named in the above -march are single and double floating point support).

I've made this change:

```
diff --git a/soc/riscv32/openisa_rv32m1/CMakeLists.txt b/soc/riscv32/openisa_rv32m1/CMakeLists.txt
index e4b5a8d535..433ac77751 100644
--- a/soc/riscv32/openisa_rv32m1/CMakeLists.txt
+++ b/soc/riscv32/openisa_rv32m1/CMakeLists.txt
@@ -3,7 +3,11 @@
 # SPDX-License-Identifier: Apache-2.0
 
 if(CONFIG_SOC_OPENISA_RV32M1_RI5CY)
-  zephyr_compile_options(-march=rv32imcxpulpv2)
+  if (CONFIG_RISCV_GENERIC_TOOLCHAIN)
+    zephyr_compile_options(-march=rv32imc)
+  else()
+    zephyr_compile_options(-march=rv32imcxpulpv2)
+  endif()
 elseif(CONFIG_SOC_OPENISA_RV32M1_ZERO_RISCY)
   zephyr_compile_options(-march=rv32imc)
 endif()
```

That only uses the extensions specifically documented to be available. Since RISCV_GENERIC_TOOLCHAIN defaults to n for this soc (see soc/riscv32/openisa_rv32m1/Kconfig.defconfig), the pulp toolchain is chosen by default. You can set RISCV_GENERIC_TOOLCHAIN to y with a conf overlay to test out the standard toolchains if you want -- no vega-specific documentation needed here as this kconfig option is already part of zephyr.

**Comment by alistair23** [comment]
> @dbkinder sorry, neglected to mention, I've addressed your comments.
> 
> @alistair23:
> 
> > With this diff I am able to build it using upstream GCC (not yet tested on hardware), if this could at least be described somewhere that would help others.
> 
> That's not really in my purview. I'm just a contractor and the recommendation I got from OpenISA is to use their toolchain. At any rate, I don't believe your patch is correct; specifically this part:
> 
> ```
>   zephyr_compile_options(-march=rv32imfdc)
> ```
> 
> If you check the reference manual, the core is an IMC core, but it's not documented to have any floating point support (the F and D extensions named in the above -march are single and double floating point support).

You are right about the floating point. I just skimmed the documentation and saw that floating point support is optional and assumed that it was included. Thanks for catching that.

> 
> I've made this change:
> 
> ```
> diff --git a/soc/riscv32/openisa_rv32m1/CMakeLists.txt b/soc/riscv32/openisa_rv32m1/CMakeLists.txt
> index e4b5a8d535..433ac77751 100644
> --- a/soc/riscv32/openisa_rv32m1/CMakeLists.txt
> +++ b/soc/riscv32/openisa_rv32m1/CMakeLists.txt
> @@ -3,7 +3,11 @@
>  # SPDX-License-Identifier: Apache-2.0
>  
>  if(CONFIG_SOC_OPENISA_RV32M1_RI5CY)
> -  zephyr_compile_options(-march=rv32imcxpulpv2)
> +  if (CONFIG_RISCV_GENERIC_TOOLCHAIN)
> +    zephyr_compile_options(-march=rv32imc)
> +  else()
> +    zephyr_compile_options(-march=rv32imcxpulpv2)
> +  endif()
>  elseif(CONFIG_SOC_OPENISA_RV32M1_ZERO_RISCY)
>    zephyr_compile_options(-march=rv32imc)
>  endif()
> ```

This looks much better :+1: 

> 
> That only uses the extensions specifically documented to be available. Since RISCV_GENERIC_TOOLCHAIN defaults to n for this soc (see soc/riscv32/openisa_rv32m1/Kconfig.defconfig), the pulp toolchain is chosen by default. You can set RISCV_GENERIC_TOOLCHAIN to y with a conf overlay to test out the standard toolchains if you want -- no vega-specific documentation needed here as this kconfig option is already part of zephyr.

Ok, sounds good to me. I'm unable to test it on hardware but at least being able to build Zephyr with standard GCC is a good start.

**Review by dbkinder — APPROVED** [review]
doc changes LGTM, thanks.

**Comment by mbolivar** [comment]
recheck

**Comment by mbolivar** [comment]
retest

**Comment by mbolivar** [comment]
Thanks for all the reviews, everyone!

The checkpatch issues should be ignored (its recommendations would break correctness), so it's just a license review now.

I've added the TSC label and asked for their review on the vendor HAL.

**Review by rgundi — COMMENTED** [review]
The changes made to the multiple interrupt levels code look good. Neatly done. Thanks for doing this. This can be considered as my approval for this section of the PR.

**Comment by galak** [comment]
Would be nice if one of the RISCV guys reviewed & approved before we merge.

**Comment by galak** [comment]
I can't remember but is this one of those ports that we want the new RISC-V toolchain for?

**Comment by mbolivar** [comment]
> Would be nice if one of the RISCV guys reviewed & approved before we merge.

@kgugala @pgielda @nategraff-sifive mind giving this a review then? You're listed in CODEOWNERS for {arch,soc,boards}/riscv32

**Review by nashif — APPROVED** [review]

**Comment by nashif** [comment]
@mbolivar mind rebasing?

**Comment by mbolivar** [comment]
> @kgugala @pgielda @nategraff-sifive mind giving this a review then? You're listed in CODEOWNERS for {arch,soc,boards}/riscv32

Ping? Sorry to bug you, but I'm under pressure to get this merged...

**Comment by mbolivar** [comment]
Hah, never mind :)

**Pull request merged**
